### PR TITLE
Persist the first-boot hostname if not otherwise set

### PR DESCRIPTION
**Pull request by darkmuggle**
This resolves the problem where `ip=...` defines a hostname but the hostname is not persisted.

### CODE REVIEW

**Comment by cgwalters** [comment]
> Is this service supposed to be CoreOS specific? If not I'd suggest renaming coreos-persist-hostname.service to ignition-persist-hostname.service..

The `spec2x` branch is currently really RHCOS specific, as is this "persist kargs into ifcfg files".  The vision for FCOS is that we use NetworkManager in the initrd, which would solve this problem.

We already have another unit with the `coreos-` naming here; so it has precedent.  (But it could be `rhcos-` I guess)

File: dracut/30ignition/coreos-persist-hostname.service [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Set the first-boot hostname as the permanent hostname.
+
+[Unit]
+Description=Persist firstboot hostname
+DefaultDependencies=false
+
+# Only trigger when the hostname is NOT A default.
+ConditionHost=!localhost
+ConditionHost=!localhost.localdomain
```
Comment (cgwalters) [inline dracut/30ignition/coreos-persist-hostname.service]: Hmm why a separate unit versus doing this as part of the existing logic for `coreos-teardown-initramfs-network.service` ?

Which currently only triggers effectively when the kernel cmdline has a non-default `ip=` - i.e. basically a static IP.

Comment (darkmuggle) [reply dracut/30ignition/coreos-persist-hostname.service]: The separate unit made the logic simplifier, e.g. `ConditionHost=localhost` and because this also applies to the DHCP situation. For RHCOS we don't want transient names to be used -- the firstboot hostname should be persisted. This ensures that.

Comment (cgwalters) [reply dracut/30ignition/coreos-persist-hostname.service]: Are we really sure this applies to DHCP too?  That scopes this code into running in e.g. AWS/GCP etc. as well as the baremetal environments that use DHCP.

Now, I don't think it's at all supported to change the hostnames returned from DHCP for OpenShift/Kubernetes today.  So, it's probably not going to *hurt* to persist it and thereafter ignore what DHCP says, but...no one has been asking us to do that either.

One thing to bear in mind here is that Kubernetes also requires the hostname to be *routable* - which implies it has to be correct in the infrastructure and not *just* the OS.  Hence, source-of-truth is the infrastructure.

The way I'd been thinking of this is that the argument for why this is RHCOS specific is that in the future in FCOS (and ideally RHCOS) with NM-in-initrd, it would parse the kernel arguments and handle this itself, correctly transferring state across the initrd.

**Comment by cgwalters** [comment]
> Finally, the use of transient hostnames is an anti-pattern for Ignition.
Changing an aspect of the machines identity (e.g. the hostname) after
provisioning can break the service running on-top the host.

I see the argument here, but the counter I think is:

- Anyone who wants to persistently set the hostname today can do so by simply writing `/etc/hostname` into Ignition
- Not all applications are going to break when the hostname changes - I mean, my desktop doesn't.  The "point of no return" here is around when a node has joined the cluster, but that is currently distinct from the Ignition boot for us.  (So we could have something on the host that screams loudly if the hostname returned from DHCP differs from the one in the kubeconfig or so?)

File: dracut/30ignition/coreos-persist-hostname.service (hunk 1 above)
Comment (darkmuggle) [reply dracut/30ignition/coreos-persist-hostname.service]: > Are we really sure this applies to DHCP too?

Better question: do we _want_ to allow changing the host via DHCP for OCP?

**Comment by dcbw** [comment]
> Finally, the use of transient hostnames is an anti-pattern for Ignition.
> Changing an aspect of the machines identity (e.g. the hostname) after
> provisioning can break the service running on-top the host.

@darkmuggle I think that Ignition requirement is in direct conflict with expectations in the DHCP and DNS cases. Here the source of truth is external to the host, and for OpenShift we must have the host agree with the external source of truth about the hostname. If the hostname is locked on install, and the external source of truth changes, we have a problem.

For DHCP your hostname may change any time the lease is renewed or re-acquired. In practice this doesn't happen often but is certainly a possibility. Same for DNS; that's entirely out of control of the host.

For OpenShift, if a node has a hostname out of sync with DHCP/DNS, kubelet will post that hostname as the node name, and other components in the cluster will try to resolve that hostname and fail. The workaround is to use IP addresses but that requires specific configuration on each node and is pretty ugly.

TLDR; my uninformed opinion is that it should work as such:
- an ignition payload for /etc/hostname: persist this hostname because admin presumably knows what they are doing
- ip=... kargs to explicitly set it: persist this hostname because admin presumably knows what they are doing
- Hostname from DHCP: use this hostname until reboot
- DNS PTR record lookup: use this hostname until reboot

**Comment by dcbw** [comment]
>     * I mean, my desktop doesn't.

@cgwalters we worked hard to make that happen like 10 years ago, IIRC we needed to make xauth not care about hostname by hardcoding localhost or something like that.

File: dracut/30ignition/coreos-persist-hostname.service (hunk 1 above)
Comment (cgwalters) [reply dracut/30ignition/coreos-persist-hostname.service]: I think we all agree that doesn't work.  The question is more - does fixing the hostname for DHCP *help* anything either?  

Now, I know for sure that if we had shipped this patch from the start it would have completely avoided the need for the [kubelet waiting on NM](https://github.com/openshift/machine-config-operator/pull/1206) patch.  But, we are on track to ship that patch long before this one.

**Comment by cgwalters** [comment]
To be 100% clear I am not *opposed* to shipping this as is.  I just would feel a lot more comfortable scoping things to only fixing the reported bug (static IP on kernel cmdline).  It's a lot simpler to explain from an architectural PoV.

**Comment by darkmuggle** [comment]
Okay, I've repushed this to deal just with the `ip=...` use case. 

The rationale for dropping the holistic approach makes sense.  Reducing this to scope of asserted identity to `ip=` cases make sense; changing the world can come later.

The problem that I have with the use of transient names for OCP is simple: it breaks certificates and makes node identity external to the cluster. One of the selling points of OCP is that the cluster is self-driving, but we make the node identity external to the cluster.

**Comment by cgwalters** [comment]
> One of the selling points of OCP is that the cluster is self-driving, but we make the node identity external to the cluster.

Yeah.  I've mentioned a few times I want to support tying identity to TPM2 devices, but that isn't universal and will never be.  We also have the core problem that the control plane is provisioned via Terraform and we don't have mechanisms yet to "adopt" that into being owned by machineAPI (where applicable).  And further, there will always be non-machineAPI platforms (manual bare metal etc.) where there just aren't APIs to control the infrastructure.

So this doesn't seem likely to change really ever, but I am hopeful we can say in *some* cases we support doing something better.

File: dracut/30ignition/persist-ifcfg.sh [hunk 2]
```diff
@@ -33,6 +33,16 @@ persist_ifcfg() {
         return 0
     fi
 
+    # Only persist the hostname if something else has not populated
+    # /etc/hostname, such as Ignition Files.
+    hname="$(< /proc/sys/kernel/hostname)"
+    hpath="/sysroot/etc/hostname"
+    if [ ! -e "${hpath}" ] && [ -n "${hname}" ]; then
+        echo -ne "${hname}" > "${hpath}"
```
Comment (cgwalters) [inline dracut/30ignition/persist-ifcfg.sh]: The `-e` isn't necessary here right?

It also looks to me like at least systemd does emit a newline character for the hostname - generally friendlier to shells and scripts to have a trailing newline.

So let's drop the `-ne`?

Comment (darkmuggle) [reply dracut/30ignition/persist-ifcfg.sh]: I added `-ne` for two reasons:
1. `systemd` does include the line, but other tooling does not. For example: ```
podman run -it --volume=/:/host:ro --cap-drop=AUDIT_WRITE,MKNOD,NET_RAW --env=TERM=xterm-256color --rm=true bash
bash-5.0# cat /etc/hostname
6b5dc518a2a0bash-5.0#
```
2. Style: I cargo culted it.

File: dracut/30ignition/persist-ifcfg.sh [hunk 3]
```diff
@@ -33,6 +33,16 @@ persist_ifcfg() {
         return 0
     fi
 
+    # Only persist the hostname if something else has not populated
+    # /etc/hostname, such as Ignition Files.
+    hname="$(< /proc/sys/kernel/hostname)"
+    hpath="/sysroot/etc/hostname"
+    if [ ! -e "${hpath}" ] && [ -n "${hname}" ]; then
```
[the remark this answers is not in the document]
Comment (cgwalters) [reply dracut/30ignition/persist-ifcfg.sh]: Ah, good catch yes, we don't want to set the persistent hostname if it's localhost.

The whole "which interface provided a hostname" is actually a mess, I bet in NetworkManager today it's just "first one wins".

(The topic of static setups but still having DHCP enabled on lingering other devices is touched on in [this bug](https://bugzilla.redhat.com/show_bug.cgi?id=1800900))

[the remark this answers is not in the document]
Comment (darkmuggle) [reply dracut/30ignition/persist-ifcfg.sh]: I'll update it to catch that case. The prior version used a check on that but that was dropped because DHCP persistent was deemed undesirable. 
```
ConditionHost=!localhost
ConditionHost=!localhost.localdomain
```

[the remark this answers is not in the document]
Comment (cgwalters) [reply dracut/30ignition/persist-ifcfg.sh]: Most elegant I think would be if dracut wrote `/etc/hostname` in the initrd only for persistent cases, then we could just mirror that.  I don't know why it isn't; probably just an oversight.

[the remark this answers is not in the document]
Comment (darkmuggle) [reply dracut/30ignition/persist-ifcfg.sh]: I suspect that the reason why dracut does not write `/etc/hostname` is because it doesn't know where the hostname comes from: it uses a udev rule to configure the device when it appears. The last device to come is used to set the hostname. The deeper I look at it, the more I wonder why there isn't a `hostname=` karg.

[the remark this answers is not in the document]
Comment (darkmuggle) [reply dracut/30ignition/persist-ifcfg.sh]: > One thing to note is that systemd will set a default transient hostname at startup, so `[ -n "${hname}" ]` will never actually fail.
> 
> IOW, this changes the behaviour subtly in the case where you set a static IP via `ip=` but _not_ a hostname: before, we'd boot all the way without any `/etc/hostname`, now we'll boot with it set to e.g. `localhost`.

Actually, no. I tested that use-case. The parts of systemd that would do that is not present in the initramfs at this point. You need `systemd-hostnamed` which is a socket listener. I evaluated that path in the last commit. Further, in testing, I confirmed that `/etc/hostname` was missing booting without an `ip=` provided hostname.

**Comment by darkmuggle** [comment]
Okay, re-pushed with handling for all the weird edge-cases.

File: dracut/30ignition/persist-ifcfg.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (jlebon) [reply dracut/30ignition/persist-ifcfg.sh]: > Actually, no. I tested that use-case. The parts of systemd that would do that is not present in the initramfs at this point.

Hmm, weird. It's basically the very first thing systemd does on both FCOS and RHCOS:

```
Welcome to Fedora CoreOS 31.20200306.dev.5 dracut-049-27.git20181204.fc31.1 (Initramfs)!

[    2.142750] systemd[1]: No hostname configured.
[    2.144456] systemd[1]: Set hostname to <localhost>.
```

And checking in `rd.break`:

```
switch_root:/# cat /proc/sys/kernel/hostname
localhost
```

File: dracut/30ignition/persist-ifcfg.sh [hunk 4]
```diff
@@ -33,6 +35,32 @@ persist_ifcfg() {
         return 0
     fi
 
+    # Persist the hostname if the admin has elected to use the dracut method of
+    # defining the IP on the commandline.
+    hpath="/sysroot/etc/hostname"
+    hname=$(< /proc/sys/kernel/hostname)
+    for iface in $(ls /sys/class/net)
```
Comment (jlebon) [inline dracut/30ignition/persist-ifcfg.sh]: Need to skip this loop if `hpath` exists already

Comment (darkmuggle) [reply dracut/30ignition/persist-ifcfg.sh]: Fixed up.

File: dracut/30ignition/persist-ifcfg.sh [hunk 5]
```diff
@@ -33,6 +35,32 @@ persist_ifcfg() {
...
+        # echo hostname > /proc/sys/kernel/hostname.
+        read _ iface_hostname _ < "${iface_totems}.hostname"
+
+        # Ensure that the hostname used by the kernel is the same one
+        # used by this interface. This guard is intended to protect against
+        # mixed `ip=` or where one is set via dhcp.
+        if [ "${iface_hostname}" == "${hname}" ] && [ ! -f "${hpath}" ]; then
+            echo "${hname}" > "${hpath}"
+            echo -ne '/etc/hostname/\0' >> /sysroot/etc/selinux/ignition.relabel
+            info "persisting hostname set by kargs for ${iface}"
```
Comment (cgwalters) [inline dracut/30ignition/persist-ifcfg.sh]:
```suggestion
            info "persisting hostname '${hname}' set by kargs for ${iface}"
```

File: dracut/30ignition/persist-ifcfg.sh [hunk 6]
```diff
@@ -33,6 +35,32 @@ persist_ifcfg() {
...
+        [ -f "${iface_totems}.hostname" ] || continue;
+
+        # The format of the file is _the command used to set the hostname_
+        # echo hostname > /proc/sys/kernel/hostname.
+        read _ iface_hostname _ < "${iface_totems}.hostname"
+
+        # Ensure that the hostname used by the kernel is the same one
+        # used by this interface. This guard is intended to protect against
+        # mixed `ip=` or where one is set via dhcp.
+        if [ "${iface_hostname}" == "${hname}" ] && [ ! -f "${hpath}" ]; then
```
Comment (jlebon) [inline dracut/30ignition/persist-ifcfg.sh]: Hmm, would be cleaner to do the `[ ! -f "${hpath}" ]` check upfront so we can avoid entering the loop entirely, but not a blocker!

Comment (darkmuggle) [reply dracut/30ignition/persist-ifcfg.sh]: Yeah, I was trying to avoid the nested if's, and figured the cost of the loop was trivial. I agree, though.
